src/kafka_handle_base.cpp

```diff
@@ -48,7 +48,7 @@ namespace cppkafka {
 const milliseconds KafkaHandleBase::DEFAULT_TIMEOUT{1000};
 
 KafkaHandleBase::KafkaHandleBase(Configuration config) 
-: timeout_ms_(DEFAULT_TIMEOUT), config_(move(config)), handle_(nullptr, nullptr) {
+: timeout_ms_(DEFAULT_TIMEOUT), config_(move(config)), handle_(nullptr, handle_deleter(this)) {
```

Comment:
keep `nullptr` here. Otherwise `rd_kafka_delete_flags` will be called with null handle and application will crash.

Reply:
That doesn't work:
```
error: no matching constructor for initialization of 'cppkafka::KafkaHandleBase::HandlePtr' (aka 'unique_ptr<rd_kafka_s, cppkafka::KafkaHandleBase::HandleDeleter>')
[build] : timeout_ms_(DEFAULT_TIMEOUT), config_(move(config)), handle_(nullptr, nullptr), destroy_flags_(0) {
```

https://stackoverflow.com/questions/50201967/why-unique-ptr-with-custom-deleter-wont-work-for-nullptr-while-shared-ptr-does

I think it's quite safe to use this here.

Reply:
Ah yes, correct, it's changed to a functor instead of a func ptr. Ok but put a check inside operator() for the handle being null.

Reply:
>Otherwise rd_kafka_delete_flags will be called with null handle and application will crash.

unique_ptr will not call deleter if there is no object, or I'm missing something?

Reply:
Seems to say that windows does not behave properly.

Reply:
That change seems a bit problematic, and I need to switch to other tasks. 
I'm totally ok if you will improve that.

Reply:
> > Otherwise rd_kafka_delete_flags will be called with null handle and application will crash.
> 
> unique_ptr will not call deleter if there is no object, or I'm missing something?
The link posted seems to imply that on Windows, the unique_ptr is not implemented properly.

Reply:
You're looking on the question in the linked SO thread, check the answer.